### PR TITLE
Made map, filter and merge in broadcast pipe

### PR DESCRIPTION
**Pull request by johnwu93**
Thank you for contributing to Coda! Please see `CONTRIBUTING.md` if you haven't
yet.

Explain your changes here.

Explain how you tested your changes here.

Checklist:

- [ ] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

Closes #0000
Closes #0000

### CODE REVIEW

**Comment by bkase** [comment]
Should we hold off reviewing this until you integrate Incremental a bit?

File: src/lib/pipe_lib/broadcast_pipe.ml [hunk 1]
```diff
@@ -94,13 +84,122 @@ module Writer = struct
...
         Deferred.unit )
 
-  let close t =
+  let rec close : type a. a t -> unit =
+   fun t ->
     guard_already_closed t (fun () ->
         Pipe.close t.root_pipe ;
-        Int.Table.iter t.pipes ~f:(fun w -> Pipe.close w) ;
+        Int.Table.iter t.pipes ~f:(function
+          | Leaf writer -> Pipe.close writer
```
Comment (enolan) [inline src/lib/pipe_lib/broadcast_pipe.ml]: I'm not sure if closing an already close pipe is safe, can you check? If it's not you need to call `Pipe.is_closed` or whatever here before trying to close.

File: src/lib/pipe_lib/broadcast_pipe.ml [hunk 2]
```diff
@@ -94,13 +84,122 @@ module Writer = struct
...
+            if Ivar.is_empty filter_pipe_ivar then (
+              let pipe, _ = create a in
+              Ivar.fill filter_pipe_ivar pipe ;
+              Deferred.unit )
+            else
+              let%bind pipe = Ivar.read filter_pipe_ivar in
+              Writer.write pipe a
+          else Deferred.unit
+        in
+        Int.Table.add_exn t.pipes ~key:reader_id
```
Comment (enolan) [inline src/lib/pipe_lib/broadcast_pipe.ml]: the filtered pipe should be seeded with the current value in the parent pipe if it passes the predicate. so add `filter_write (peek t)` here

File: src/lib/pipe_lib/broadcast_pipe.ml [hunk 3]
```diff
@@ -94,13 +84,122 @@ module Writer = struct
...
+             | Node {write; _} -> write v )
+         in
+         Pipe.Consumer.values_sent_downstream consumer ;
+         let%bind () =
+           Deferred.List.iter ~how:`Parallel inner_pipes
+             ~f:(function
+             | Leaf writer -> Deferred.ignore @@ Pipe.downstream_flushed writer
+             | Node _ ->
+                 Deferred.unit
+                 (* A Broadcast pipe would get it's write determined after all of it's readers are consumed *) )
```
Comment (enolan) [inline src/lib/pipe_lib/broadcast_pipe.ml]: This comment is confusing and I was sure you had done it wrong until I read further. I see now that we're relying on the `write` function inside the `Node` blocking until the downstream consumers are done. Can you make it clear that `write` always calls `Broadcast_pipe.Writer.write`?

File: src/lib/pipe_lib/broadcast_pipe.ml [hunk 4]
```diff
@@ -94,13 +84,122 @@ module Writer = struct
...
+         Ivar.fill !downstream_flushed_v () ;
+         Deferred.unit )) ;
+  (t, t)
+
+module Reader = struct
+  include Reader0
+
+  (* The only reason for a descendant pipe to get removed from it's parent if it closes.
+      The only way to do that right now is to close the pipe from the parent.
+      Therefore, we do not have to be concerned about removing a mapped pipe. *)
```
Comment (enolan) [inline src/lib/pipe_lib/broadcast_pipe.ml]: if you call `to_jane_street_pipe` on the result of `map` you can close the descendant.

File: src/lib/pipe_lib/broadcast_pipe.ml [hunk 5]
```diff
@@ -142,6 +232,95 @@ let%test_unit "listeners properly receive updates" =
...
+      in
+      let d =
+        expect_pipe (module Test_input) merged_reader [`A; `B; `D; `A; `C]
+      in
+      let%bind () = Writer.write first_writer `B in
+      let%bind () = Writer.write second_writer `D in
+      let%bind () = Writer.write first_writer `A in
+      let%bind () = Writer.write second_writer `C in
+      Writer.close first_writer ;
+      (* TODO: uncommenting the below code would cause the writer to throw an `Already_closed` exception. Handle that *)
```
Comment (enolan) [inline src/lib/pipe_lib/broadcast_pipe.ml]: do the to do

**Comment by johnwu93** [comment]
We are moving away from using Broadcast_pipe for the graphql services. This seems unneeded in the short term.

**Pull request closed without merging**